### PR TITLE
[Feat] Add unified evaluation script for MoM collection

### LINKED ISSUE #1137: [v0.2-Athena]: Create unified Python evaluation script for multi-lingual MoM collection models
## Description

Create a unified Python evaluation script to assess the accuracy of all models in the new multi-lingual MoM (Mixture of Models) collection. This script should provide a standardized way to evaluate model performance locally across different task types (text classification, token classification, etc.).

## Context

The multi-lingual MoM collection on HuggingFace contains 10 specialized models:
- **Text Classification Models:**
  - `llm-semantic-router/mmbert-feedback-detector-merged` (0.3B)
  - `llm-semantic-router/mmbert-jailbreak-detector-merged` (0.3B)
  - `llm-semantic-router/mmbert-fact-check-merged` (0.3B)
  - `llm-semantic-router/mmbert-intent-classifier-merged` (0.3B)
  
- **Token Classification Models:**
  - `llm-semantic-router/mmbert-pii-detector-merged` (0.3B)

- **LoRA Variants:**
  - `llm-semantic-router/mmbert-feedback-detector-lora`
  - `llm-semantic-router/mmbert-jailbreak-detector-lora`
  - `llm-semantic-router/mmbert-fact-check-lora`
  - `llm-semantic-router/mmbert-intent-classifier-lora`
  - `llm-semantic-router/mmbert-pii-detector-lora`

Currently, the codebase has evaluation scripts for reasoning tasks (MMLU-Pro, ARC Challenge) but lacks a unified evaluation framework for these classification models.

## Requirements

The evaluation script should:

1. **Support Multiple Model Types:**
   - Text classification (intent, jailbreak, fact-check, feedback)
   - Token classification (PII detection)
   - Both merged and LoRA variants

2. **Provide Comprehensive Metrics:**
   - Overall accuracy
   - Per-class precision, recall, F1-score
   - Confusion matrix
   - Inference latency (avg, p50, p99)

3. **Support Multiple Datasets:**
   - Use appropriate test datasets for each model type
   - Support custom dataset loading
   - Handle multilingual test cases

4. **Output Structured Results:**
   - JSON format for programmatic access
   - Human-readable summary reports
   - Per-model result files in organized directory structure

5. **Configuration Options:**
   - Model selection (single or multiple models)
   - Batch size configuration
   - Sample size limits for quick testing
   - Device selection (CPU/GPU)

## Suggested Implementation

Create `src/training/model_eval/mom_collection_eval.py` with:

```python
# Key features:
- Model registry mapping HF model IDs to task types
- Task-specific evaluation functions
- Unified metrics calculation
- Parallel evaluation support
- Progress tracking with tqdm
```

## Reference

- HuggingFace Collection: https://huggingface.co/collections/llm-semantic-router/multi-lingual
- Existing evaluation scripts: `src/training/model_eval/mmlu_pro_vllm_eval.py`, `src/training/model_eval/arc_challenge_vllm_eval.py`
- Benchmark framework: `bench/hallucination/evaluate.py`

## Acceptance Criteria

- [ ] Script can evaluate all 10 models in the collection
- [ ] Supports both local model files and HuggingFace Hub loading
- [ ] Generates accuracy metrics for each model
- [ ] Outputs results in JSON and human-readable formats
- [ ] Includes usage examples in documentation
- [ ] Can run on CPU and GPU
- [ ] Handles errors gracefully (missing models, dataset issues)

### PR DESCRIPTION
**Pull request by ZohaibHassan16**
FIX #1137 

I have created a unified python script to evaluate the accuracy of the new MoM (Mixture of Models) collection.

**What I did:**
- Added `src/training/model_eval/mom_collection_eval.py` which can evaluate all 10 models(Intent, Jailbreak, Feedback, Fact-check, and PII).
- The script handles both **Text Classification** and **Token Classification** tasks.
- **LoRA Support:** Supports merged models and LoRA variants (using the `--use_lora` flag).
- **Local Model Support: ** Added `--model_id` to evaluate local fine-tuned checkpoints or custom paths.
- **Performance:** Implemented **batch processing** `--batch_size` for faster inference on CPU/GPU.
- **Documentation:** Added `README.md` with detailed usage examples.
- **Dependencies:** Updated `requirements.txt` with necessary libraries (`peft`, `seqeval`, etc.).

**How it works:**
- Calculates metrics like Accuracy, F1-score, Precision and Recall.
- **Robustness:** if the real dataset isn't public on HuggingFace yet, the script automatically detects this and falls back to dummy data instead of crashing.

**Verification:** I tested this locally on my CPU. As shown in the screenshot below, the script successfully:
1. Downloads the model.
2. Handles the missing dataset warning gracefully.
3. Runs inference using **batch processing** (Batch size: 2).
4. Generates the JSON report and Confusion Matrix.

Screenshot 2026-01-22 032832

### CODE REVIEW

**Comment by ZohaibHassan16** [comment]
All CI checks are now passing. Ready for review !! @Xunzhuo @wangchen615 @rootfs

**Comment by Xunzhuo** [comment]
can you update all other models eval results?

**Comment by Xunzhuo** [comment]
for pii and intent:
https://github.com/vllm-project/semantic-router/blob/main/src/training/training_lora/classifier_model_fine_tuning_lora/ft_linear_lora.py#L136
https://github.com/vllm-project/semantic-router/blob/main/src/training/training_lora/pii_model_fine_tuning_lora/pii_bert_finetuning_lora.py#L175
for jailbreak and feedback
https://huggingface.co/datasets/llm-semantic-router/jailbreak-detection-dataset
https://huggingface.co/datasets/llm-semantic-router/feedback-detector-dataset

**Comment by ZohaibHassan16** [comment]
Thanks for the feedback!! @Xunzhuo 
I will update the script to use the actual datasets you referenced and I'll also run the evaluation locally on these real datasets and update the PR to that effect.
Working on it now

**Comment by ZohaibHassan16** [comment]
Thanks for the feedback!! @Xunzhuo 
I will update the script to use the actual datasets you referenced and I'll also run the evaluation locally on these real datasets and update the PR to that effect.
Working on it now

**Comment by Xunzhuo** [comment]
fact check: https://huggingface.co/datasets/llm-semantic-router/fact-check-classification-dataset

**Comment by szedan-rh** [comment]
Can you please make the constant variables in constant.py and import that to the script much more cleaner the code will be

**Comment by ZohaibHassan16** [comment]
Thanks for the feedback @szedan-rh , sure thing.

**Comment by ZohaibHassan16** [comment]
Hi @Xunzhuo ,

I’ve completely overhauled the evaluation script !! I looked into the training scripts for each model as you suggested and integrated real datasets.

Here’s the breakdown of what’s new:

**Specialized Dataset Handling:** I’ve implemented specific logic to handle every dataset type in the MoM registry:

 - Intent: Specialized filtering for MMLU-Pro.

  - PII: Automatic fetching and processing of the Microsoft Presidio dataset.

  - Standard Classification: Auto-detection logic for Jailbreak, Feedback, and Fact-check datasets 

**Performance**: I added batch processing and parallel evaluation support so it's much faster, even on a CPU.

**Multilingual Support:** I added a `--language` flag with specialized logic (using character ranges) to filter datasets for Spanish, Arabic, etc. 

**Robustness**: Added OOM recovery and logic to handle cases where a dataset might not be publicly reachable yet.


**Local Verification**: I ran this locally on my machine, and it’s working great. I've attached screenshots showing the final summary table. It also saves full metrics to a JSON report.

**Parellel testing:**

Screenshot 2026-01-22 133502

**Multilingual testing:**

Screenshot 2026-01-22 133622

File: src/training/model_eval/mom_collection_eval.py [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+            logger.info(f"Loading custom dataset: {args.custom_dataset}")
+            ext = Path(args.custom_dataset).suffix
+            if ext not in [".json", ".csv"]:
+                raise ValueError(f"Unsupported format: {ext}. Use .json or .csv")
+
+            ds = retry_operation(
+                lambda: load_dataset(
+                    "json" if ext == ".json" else "csv",
+                    data_files=args.custom_dataset,
+                    split="train",
```
Comment (rootfs) [inline src/training/model_eval/mom_collection_eval.py]: use eval or test first and fall back to train to make test using unseen data.

**Comment by ZohaibHassan16** [comment]
Hey @rootfs , I've implemented the change as you said:
- Custom dataset now prioritizes `"test"` split first
- Falls back to `"validation"` / `"eval"`
- Falls back to `"train"` only as last resort
- Added clear logging of which split was loaded
- Added a warning when falling back to train (to alert users about possible optimistic results / overfitting)

**Comment by Xunzhuo** [comment]
hey @rootfs can you take a look again : )
